app/assets/javascripts/shared/datatable/main.js

```diff
@@ -169,11 +178,47 @@ class DradisDatatable {
     return ids;
   }
 
+  // The default data returned by DataTables doesn't contain any information to help us
+  // identify which column it originated from, so we modify it before saving
+  // in localStorage.
+  saveHeadersInData(data) {
+    this.tableHeaders.forEach(function(th, index) {
+      data.columns[index].header = th.textContent;
+    })
+
+    return data;
+  }
+
   unbindDataTable() {
     var that = this;
 
     document.addEventListener('turbolinks:before-cache', function() {
       that.dataTable.destroy();
     });
   }
+
+  // This method remove old columns from the data stored in localStorage
+  // (indirectly, because we are iterating over table headers on the page)
+  // and add new columns to it
+  rebuildLocalStorageColumns(localStorageData) {
+    var newColumns = [];
+
+    this.tableHeaders.forEach(function(th, index) {
+      // Check if there's a key inside the column object with the same value as
+      // the table header text
+      var column = localStorageData.columns.find(function(column) {
+        return column.header == th.textContent;
+      })
+
+      if (column) {
+        newColumns.push(column);
+      } else {
+        // Add new column with default visibility = false
+        newColumns.push({ visible: false });
+      }
+    })
```

Comment:
there's something about this loop within a loop to look for a single item that bothers me...

In pseudocode:
```js
var columnData = { visible: false };

localStorageData.columns.find(function(column) {
  if (column.header == th.textContent) {
    columnData = column;
    break;
  }
});
newColumns.push(columnData);
```

Reply:
It's not looking for _just a_ single item.

It's looking for a column that matches the current iterator's text value, if found, we re-use the column stored in localStorage as it might contain the column's state.

Javascript's `Array.prototype.find()` returns a single value if found, and null otherwise, so we don't have to break it.